contracts/starknet/jug.cairo

```diff
@@ -0,0 +1,433 @@
+@event
+func Drip(ilk: felt) {
+}
+
+// modifier auth {
+//         require(wards[msg.sender] == 1, "Jug/not-authorized");
+//         _;
+//     }
+func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
+    let (caller) = get_caller_address();
+    let (ward) = _wards.read(caller);
+    with_attr error_message("Jug/not-authorized") {
+        assert ward = 1;
+    }
+    return ();
+}
+
+@view
+func ilks{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(i: felt) -> (ilk: Ilk) {
+    let (res) = _ilks.read(i);
+    return (res,);
+}
+
+@view
+func wards{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(user: felt) -> (
+    res: felt
+) {
+    let (res) = _wards.read(user);
+    return (res,);
+}
+
+@view
+func vat{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (res: felt) {
+    let (res) = _vat.read();
+    return (res,);
+}
+
+// // --- Init ---
+//     constructor(address vat_) {
+//         wards[msg.sender] = 1;
+//         vat = VatLike(vat_);
+//         emit Rely(msg.sender);
+//     }
+@constructor
+func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    vat_: felt, ward_: felt
+) {
+    _wards.write(ward_, 1);
+
+    _vat.write(vat_);
+
+    Rely.emit(ward_);
+
+    return ();
+}
+
+// // --- Math ---
+//     function _rpow(uint256 x, uint256 n, uint256 b) internal pure returns (uint256 z) {
+//       assembly {
+//         switch x case 0 {switch n case 0 {z := b} default {z := 0}}
+//         default {
+//           switch mod(n, 2) case 0 { z := b } default { z := x }
+//           let half := div(b, 2)  // for rounding.
+//           for { n := div(n, 2) } n { n := div(n,2) } {
+//             let xx := mul(x, x)
+//             if iszero(eq(div(xx, x), x)) { revert(0,0) }
+//             let xxRound := add(xx, half)
+//             if lt(xxRound, xx) { revert(0,0) }
+//             x := div(xxRound, b)
+//             if mod(n,2) {
+//               let zx := mul(z, x)
+//               if and(iszero(iszero(x)), iszero(eq(div(zx, x), z))) { revert(0,0) }
+//               let zxRound := add(zx, half)
+//               if lt(zxRound, zx) { revert(0,0) }
+//               z := div(zxRound, b)
+//             }
+//           }
+//         }
+//       }
+//     }
+func _rpow{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*
+}(x: Uint256, n: Uint256, base: Uint256) -> (z: Uint256) {
+    alloc_locals;
+    let (is_x_null) = uint256_eq(x, Uint256(0, 0));
+    let (is_n_null) = uint256_eq(n, Uint256(0, 0));
+
+    if (is_x_null == 1) {
+        if (is_n_null == 1) {
+            return (z=base);
+        } else {
+            return (z=Uint256(0, 0));
+        }
+    } else {
+        let (_, mod_2_n) = div_rem(n, Uint256(2, 0));
+        let (is_mod_null) = uint256_eq(mod_2_n, Uint256(0, 0));
+        let (_z) = internal.get_z(base, x, is_mod_null);
```

Comment:
I would inline `get_z`.

Reply:
Not ideal as we'd have to use `tempvar` and we will have weird implicit argument conflicts